pkg/controllers/image/image.go

```diff
@@ -0,0 +1,338 @@
+	}
+
+	switch secret.Type {
+	case corev1.SecretTypeBasicAuth:
+		return &http.BasicAuth{
+			Username: string(secret.Data[corev1.BasicAuthUsernameKey]),
+			Password: string(secret.Data[corev1.BasicAuthPasswordKey]),
+		}, nil
+	case corev1.SecretTypeSSHAuth:
+		publicKey, err := ssh.NewPublicKeys("git", secret.Data[corev1.SSHAuthPrivateKey], "")
+		if err != nil {
+			return nil, err
+		}
+		return publicKey, nil
+	}
+	return nil, errors.New("invalid secret type")
+}
+
+// authFromSecret creates an Authenticator that can be given to the
+// `remote` funcs, from a Kubernetes secret. If the secret doesn't
+// have the right format or data, it returns an error.
+func authFromSecret(secret *corev1.Secret, registry string) (authn.Authenticator, error) {
+	switch secret.Type {
+	case "kubernetes.io/dockerconfigjson":
+		var dockerconfig struct {
+			Auths map[string]authn.AuthConfig
+		}
+		configData := secret.Data[".dockerconfigjson"]
+		if err := json.NewDecoder(bytes.NewBuffer(configData)).Decode(&dockerconfig); err != nil {
+			return nil, err
+		}
+		auth, ok := dockerconfig.Auths[registry]
+		if !ok {
+			return nil, fmt.Errorf("auth for %q not found in secret %v", registry, types.NamespacedName{Name: secret.GetName(), Namespace: secret.GetNamespace()})
+		}
+		return authn.FromConfig(auth), nil
+	default:
+		return nil, fmt.Errorf("unknown secret type %q", secret.Type)
+	}
+}
+
+func shouldScan(image *v1alpha1.ImageScan) bool {
+	interval := image.Spec.Interval
+	if interval.Seconds() == 0.0 {
+		interval = metav1.Duration{
+			Duration: time.Minute * 5,
+		}
+	}
+	if image.Status.LastScanTags == nil {
+		return true
+	}
+
+	if time.Now().Sub(image.Status.LastScanTime.Time) < interval.Duration {
+		return false
+	}
+	return true
+}
+
+func latestTag(policy v1alpha1.ImagePolicyChoice, versions []string) (string, error) {
+	switch {
+	case policy.SemVer != nil:
+		return semverLatest(policy.SemVer.Range, versions)
+	case policy.Alphabetical != nil:
+		var des bool
+		if policy.Alphabetical.Order == "" {
+			des = true
+		} else {
+			des = policy.Alphabetical.Order == AlphabeticalOrderDesc
+		}
+		var latest string
+		for _, version := range versions {
+			if latest == "" {
+				latest = version
+				continue
+			}
+
+			if version >= latest && des {
+				latest = version
+			}
+
+			if version <= latest && !des {
+				latest = version
+			}
+		}
+		return latest, nil
+	default:
+		return semverLatest("*", versions)
+	}
+	return "", errors.New("failed to find the latest version")
+}
+
+func semverLatest(r string, versions []string) (string, error) {
+	contraints, err := semver.NewConstraint(r)
+	if err != nil {
+		return "", err
+	}
+	var latestVersion *semver.Version
+	for _, version := range versions {
+		if ver, err := semver.NewVersion(version); err == nil {
+			if latestVersion == nil || ver.GreaterThan(latestVersion) {
+				if contraints.Check(ver) {
+					latestVersion = ver
+				}
+			}
+		}
+	}
+	return latestVersion.Original(), nil
```

Comment:
Is it possible that `latestVersion` will remain uninitialized? Perhaps there should be an error returned if it's uninitialized.

Reply:
Good catch here :+1: